### PR TITLE
Add PolicyType and Policy models

### PR DESCRIPTION
**Pull request by m4dcoder**
Add appropriate DB, persistence, and API models, REST controllers, and CLI operations for PolicyType and Policy.

### CODE REVIEW

File: st2api/st2api/controllers/v1/policies.py [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+    @jsexpose(body_cls=PolicyAPI, status_code=http_client.CREATED)
+    def post(self, instance):
+        """
+            Create a new policy.
+            Handles requests:
+                POST /policies/
+        """
+        op = 'POST /policies/'
+
+        if not hasattr(instance, 'enabled'):
```
Comment (manasdk) [inline st2api/st2api/controllers/v1/policies.py]: Could be handled in DB?

File: st2api/st2api/controllers/v1/policies.py [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+            LOG.debug('%s found object: %s', op, db_model)
+        except Exception as e:
+            LOG.exception('%s unable to find object.', op)
+            abort(http_client.NOT_FOUND, e.message)
+            return
+
+        db_model_id = db_model.id
+
+        try:
+            validate_not_part_of_system_pack(db_model)
```
Comment (manasdk) [inline st2api/st2api/controllers/v1/policies.py]: Why not?

File: st2api/st2api/controllers/v1/policies.py [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+
+        return filters
+
+
+class PolicyController(resource.ContentPackResourceController):
+    model = PolicyAPI
+    access = Policy
+
+    supported_filters = {
+        'pack': 'pack'
```
Comment (manasdk) [inline st2api/st2api/controllers/v1/policies.py]: `resource_ref ` could also be a useful filter

Comment (m4dcoder) [inline st2api/st2api/controllers/v1/policies.py]: Good point!

File: st2client/st2client/client.py [hunk 4]
```diff
@@ -73,55 +73,79 @@ def __init__(self, base_url=None, auth_url=None, api_url=None, api_version=None,
 
         # Instantiate resource managers and assign appropriate API endpoint.
         self.managers = dict()
-        self.managers['Token'] = models.ResourceManager(
-            models.Token, self.endpoints['auth'], cacert=self.cacert, debug=self.debug)
-        self.managers['RunnerType'] = models.ResourceManager(
-            models.RunnerType, self.endpoints['api'], cacert=self.cacert, debug=self.debug)
+
```
Comment (manasdk) [inline st2client/st2client/client.py]: Would be nice to avoid stylistic changes if the affected files > 5.

File: st2api/st2api/controllers/v1/policies.py (hunk 2 above)
Comment (m4dcoder) [inline st2api/st2api/controllers/v1/policies.py]: I copied this from the ContentPack controller.  I guess I can use cases where users may want to either delete and change the policies that are deployed as part of the core pack.  But it is going to be inconsistent with how resources (i.e. actions) are managed under the core pack.  Are we ok with that?

File: st2api/st2api/controllers/v1/policies.py (hunk 1 above)
Comment (m4dcoder) [inline st2api/st2api/controllers/v1/policies.py]: Yes.  I just copied this from https://github.com/StackStorm/st2/blob/master/st2api/st2api/controllers/v1/actions.py#L80.

File: st2client/st2client/client.py (hunk 4 above)
Comment (m4dcoder) [inline st2client/st2client/client.py]: Ok.  I'll keep that in mind.  It is too tempting.  The stuffs are not sorted and it was hard to read.... :P

File: st2common/st2common/models/db/policy.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        help_text='A flag indicating whether the runner for this type is enabled.')
+    module = me.StringField(
+        required=True,
+        help_text='The python module that implements the policy for this type.')
+    parameters = me.DictField(
+        help_text='The specification for parameters for the policy type.')
+
+    def __init__(self, *args, **kwargs):
+        super(PolicyTypeDB, self).__init__(*args, **kwargs)
+        self.ref = PolicyTypeReference.to_string_reference(resource_type=self.resource_type,
```
Comment (manasdk) [inline st2common/st2common/models/db/policy.py]: The fact that this ref is diffrent from other ref in the system bothers me. This ref does not follows the standard `pack.name` convention either.

I see the value in scope by name of the type but couldn't that be simply captured in the name? To deviate from the usual form of ref which is basically the way to uniquely identify, in a human readable form, objects in our system we need absolute conviction. I am not convinced.

**Comment by manasdk** [comment]
I am +1 to the overall change.

I am not in favor of `PolicyTypeReference` as meaning anything other than ref i.e. `pack.name`.

File: st2common/st2common/models/db/policy.py (hunk 5 above)
Comment (m4dcoder) [inline st2common/st2common/models/db/policy.py]: I thought about that and started initially to name the policy like "action.concurrency".  But how is it going to be enforced?  Just checking the name and then throw an exception that user needs to prefix it with "action" (or other type)?

File: st2api/st2api/controllers/v1/policies.py (hunk 2 above)
Comment (manasdk) [inline st2api/st2api/controllers/v1/policies.py]: Policies are fundamentally deployment specific. Heck even the fact that `system` actions cannot be changed is broken in my view.

File: st2client/st2client/client.py (hunk 4 above)
Comment (manasdk) [inline st2client/st2client/client.py]: I understand the temptation but if you believe that you would like the code to be reviewed then keeping in mind that this adds noise and make reviewing harder is also important.

File: st2common/st2common/models/db/policy.py (hunk 5 above)
Comment (manasdk) [inline st2common/st2common/models/db/policy.py]: Why enforce?

File: st2api/st2api/controllers/resource.py [hunk 6]
```diff
@@ -188,7 +188,7 @@ def _get_by_name(self, resource_name):
         return resource_db
 
 
-class ContentPackResourceControler(ResourceController):
+class ContentPackResourceController(ResourceController):
```
Comment (lakshmi-kannan) [inline st2api/st2api/controllers/resource.py]: These changes should definitely be in its own PR.

File: st2client/st2client/client.py (hunk 4 above)
Comment (m4dcoder) [inline st2client/st2client/client.py]: Got it.

File: st2api/st2api/controllers/v1/policies.py [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+            abort(http_client.BAD_REQUEST, str(e))
+            return
+
+        LOG.debug('%s updated object: %s', op, db_model)
+        LOG.audit('Policy updated. Policy.id=%s' % (db_model.id), extra={'policy_db': db_model})
+
+        return self.model.from_model(db_model)
+
+    @jsexpose(arg_types=[str], status_code=http_client.NO_CONTENT)
+    def delete(self, ref_or_id):
```
Comment (lakshmi-kannan) [inline st2api/st2api/controllers/v1/policies.py]: I am not sure if we should support delete this early. Don't we have to delete all references to the policy before deleting the policy itself?

File: st2client/st2client/client.py (hunk 4 above)
Comment (lakshmi-kannan) [inline st2client/st2client/client.py]: Plus you are gonna see merge conflict with https://github.com/StackStorm/st2/pull/1480 :P.

File: st2common/st2common/models/db/policy.py (hunk 5 above)
Comment (m4dcoder) [inline st2common/st2common/models/db/policy.py]: > I see the value in scope by name of the type
I prefer organization and cleanliness.  Enforcing some naming standard makes the UI cleaner.  But if this is causing a lot of heartache, I can remove it.

File: st2api/st2api/controllers/v1/policies.py (hunk 7 above)
Comment (m4dcoder) [inline st2api/st2api/controllers/v1/policies.py]: Do you mean the other way around?  Policy overlays on top of a resource (i.e. action).  Deleting a policy should be independent from the action.  On the other hand, if a user deletes an action for example, should all the existing policies that reference the action be deleted as well?  At this point (experimental), I don't think it matters.  If the action is deleted, the action would never be executed, and any policies that reference the action won't be touched.

File: st2client/st2client/client.py (hunk 4 above)
Comment (m4dcoder) [inline st2client/st2client/client.py]: Hm, maybe I should merge now. :p

File: st2api/st2api/controllers/v1/policies.py (hunk 7 above)
Comment (lakshmi-kannan) [inline st2api/st2api/controllers/v1/policies.py]: I guess I don't how the association is going to established between policies and content at this point. So yeah, deleting a policy cannot be simple as just deleting the model from db. It probably has more logic to it is all I am saying. I am fine with leaving the code as such.

File: st2common/st2common/models/db/policy.py (hunk 5 above)
Comment (manasdk) [inline st2common/st2common/models/db/policy.py]: > organization and cleanliness

organization and cleanliness are subjective for the most part. Carrot and no stick would be my approach. Same property different meaning depending on the object leads to confusion. 

Anyway we are discussing 2 separate matters -
1. ref property meaning something different PolicyType. Should not.
2. Enforcing a specific structure for any naming property - that is arguable but I am not against.

File: st2api/st2api/controllers/v1/policies.py (hunk 2 above)
Comment (m4dcoder) [inline st2api/st2api/controllers/v1/policies.py]: Another question, what should the behavior be for a st2 upgrade?  If users modified/deleted the policies in core, during an upgrade, should we overlay an existing installation?

File: st2api/st2api/controllers/v1/policies.py (hunk 7 above)
Comment (m4dcoder) [inline st2api/st2api/controllers/v1/policies.py]: Sure. When we can articulate what that association is, we can add logic to the delete.

**Comment by manasdk** [comment]
I guess we all have decided not to block on PolicyTypeReference.

**Comment by m4dcoder** [comment]
Except for validate_not_part_of_system_pack(db_model), I addressed comments above. If we want to allow PUT/DELETE for policies in system pack, I can remove validate_not_part_of_system_pack(db_model) in a separate commit.
